### PR TITLE
[Feature] #6174 oss_ext read or write data from aliyun oss

### PR DESCRIPTION
**Pull request by wjzeng**
HI team

I want to submit an external table oss_ext, let greenplum user can reads and writes the data on aliyun oss.

oss_ext already open source to aliyun github https://github.com/aliyun/oss_ext

I will submit to gpdb github soon.

looking forward to the team's response.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@wjzeng Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=6175)!

[Click here](https://cla.pivotal.io/sync/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=6175) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by yydzero** [comment]
It is great to see this contribution. Thanks @wjzeng!

@liming01 @pf-qiu @xiong-gang This is external table implementation for aliyun's OSS which is similar as S3.  Would you please take a look at this PR? Thanks!

**Comment by pivotal-issuemaster** [comment]
@wjzeng Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=6175)!

**Comment by wjzeng** [comment]
I got this error message in concourse-ci/status .
It seems that the oss_ext part is not involved.
Please help me solve this problem.

----------------------------------------------------------------------
Ran 34 tests in 36.461s

FAILED (failures=1)
Makefile:2: recipe for target 'installcheck' failed
make[2]: *** [installcheck] Error 1
make[2]: Leaving directory '/tmp/build/ef6a11cc/gpdb_src/gpMgmt/bin/gpload_test'
Makefile:281: recipe for target 'installcheck' failed
make[1]: *** [installcheck] Error 2
make[1]: Leaving directory '/tmp/build/ef6a11cc/gpdb_src/gpMgmt/bin'
GNUmakefile:142: recipe for target 'installcheck-world-gpMgmt/bin-recurse' failed
make: *** [installcheck-world-gpMgmt/bin-recurse] Error 2
Traceback (most recent call last):
  File "gpdb_src/concourse/scripts/test_gpdb.py", line 95, in <module>
    sys.exit(main())
  File "gpdb_src/concourse/scripts/test_gpdb.py", line 90, in main
    copy_output()
  File "gpdb_src/concourse/scripts/test_gpdb.py", line 39, in copy_output
    shutil.copyfile("gpdb_src/src/test/regress/regression.diffs", "icg_output/regression.diffs")
  File "/usr/lib/python2.7/shutil.py", line 82, in copyfile
    with open(src, 'rb') as fsrc:
IOError: [Errno 2] No such file or directory: 'gpdb_src/src/test/regress/regression.diffs'

File: gpAux/extensions/aliyun_oss_ext/oss_ext.c [hunk 1]
```diff
@@ -0,0 +1,1325 @@
...
+		else if (myData->async == true)
+		{
+			CreateAsyncOssSource(myData);
+		}
+		else if (myData->async == false)
+		{
+			CreateOssSource(myData);
+		}
+
+		EXTPROTOCOL_SET_USER_CTX(fcinfo, myData);
```
[the remark this answers is not in the document]
Comment (wjzeng) [reply gpAux/extensions/aliyun_oss_ext/oss_ext.c]: thanks, i will fix that.

[the remark this answers is not in the document]
Comment (wjzeng) [reply gpAux/extensions/aliyun_oss_ext/oss_ext.c]: HI @pf-qiu
I do some test for test clean myData When a transaction is cancelled
and i found that myData was released when the SQL statement was canceled.
It seems that myData is released when the statement is rolled back.

The specific steps are
1 
I am add the debug code(elog(WARNING, "free_data");) like before free_data(myData);  code like this

 if (EXTPROTOCOL_IS_LAST_CALL(fcinfo))
 {
                if (myData)
                {
                        elog(WARNING, "free_data");
                        free_data(myData);
                }
                PG_RETURN_INT32(0);
}
build and install to bin dir.

2 run a long sql about oss_ext by psql
explain analyze select * from lineitem_ext;

3 run control +c to cancel this sql

4 then, i get this message

postgres=# explain analyze select * from lineitem_ext;
^CCancel request sent
NOTICE:  a total of 32 files will be loaded, asyncmode  (seg0 slice1 192.168.0.45:25432 pid=24223)
CONTEXT:  External table lineitem_ext, file oss://oss-cn-shanghai.aliyuncs.com prefix=300gb_tpch/lineitem.csv id=1SwiM8h6hJ6TAg5g key= WZ26VrrXBp8XKuR2RHn58EigSA04yR bucket=osshuadong2
WARNING:  free_data  (seg1 slice1 192.168.0.45:25433 pid=24224)
WARNING:  free_data  (seg0 slice1 192.168.0.45:25432 pid=24223)
WARNING:  free_data  (seg2 slice1 192.168.0.45:25434 pid=24225)
NOTICE:  canceling statement due to user request

It is looks like the sql are rollback, and myData is released.

**Comment by wjzeng** [comment]
> Hi, @wjzeng . Thanks for this great contribution.
> We may gradually retire `gpAux` folder, so better to put code under `gpcontrib` folder.

Thanks for your reply, I will move the location of the oss_ext

File: gpAux/extensions/aliyun_oss_ext/oss_ext.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (wjzeng) [reply gpAux/extensions/aliyun_oss_ext/oss_ext.c]: HI @pf-qiu 
i am add one callback function for clean myData.
please check my code one more time.

**Comment by schubert** [comment]
flagging @pivotal-mike to chime in here.

File: gpcontrib/aliyun_oss_ext/test/setup_data.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+osscmd config --host=$oss_host --id=$oss_id --key=$oss_key
+
+#cleanup existing dir
+osscmd deleteallobject --force=true oss://$oss_bucket/oss_reg_test/
+osscmd deleteallobject --force=true oss://$oss_bucket/oss_reg_test2/
+osscmd deleteallobject --force=true oss://$oss_bucket/cdn_demo_20170824/
+osscmd deleteallobject --force=true oss://$oss_bucket/cdn_demo_201801/
+osscmd deleteallobject --force=true oss://$oss_bucket/oss_reg_test/expdir/
+
```
Comment (liming01) [inline gpcontrib/aliyun_oss_ext/test/setup_data.sh]: It seems the data directory name on alicloud is fixed, which means only one this test can be running, which will be error prone if PR testing or commit testing running this test concurrently.
So please make this directory unique for each instance of this testing.

File: gpcontrib/aliyun_oss_ext/oss_ext.c [hunk 3]
```diff
@@ -0,0 +1,1364 @@
...
+
+	/* put into local buffer */
+	memcpy(myData->buffer, data, request_len);
+	myData->offset += request_len;
+
+	return request_len;
+}
+
+static void
+flush_and_switch_to_next_file(ext_oss_t  *myData)
```
Comment (liming01) [inline gpcontrib/aliyun_oss_ext/oss_ext.c]: Have you tested it? i.e. write to multiple files on alicloud.

**Comment by wjzeng** [comment]
@adam8157 ok, I will do that.

**Comment by wjzeng** [comment]
@liming01 The default is append mode .
The oss file that oss_ext wirte to oss , format is the table name +$timestamp_to_str. $timestamp_to_str will not be repeated, so it will not overwrite.

**Comment by pivotal-mike** [comment]
Thank you very much for your contribution of the Alicloud external table extension.  We are really excited to see this level of outside contribution to Greenplum.  Unfortunately your contribution presents some challenges that makes it not possible for us to accept into the main code base:

- The testing and development requires access to alicloud for both development and presents some challenges for our continuous integration and build system. If you have ideas or suggestions how we could work around this we’d love to know more.
- The memory management used in this extension does not leverage the PostgreSQL nor the GPDB memory management routines. Details can be found here: https://github.com/postgres/postgres/blob/REL9_4_STABLE/src/backend/utils/mmgr/README
- The Apache 2.0 license is very much appreciated however it due to the size of this contribution (14k LoC) we think it may require review by Pivotal legal.
- The code as presented is suitably large and complex and we are not in a position where we feel, give the above, we can reliably maintain this.

I’d like to propose an alternative, please let me know your thoughts on this:

As this is written as a proper extension, we’d like to suggest that we promote this on the greenplum.org site.  The code could be maintained in another github repo but it would be reachable and listed on the Open Source GPDB web site.  We have the PostGIS extension set up this way: https://github.com/greenplum-db/geospatial. This would have the added benefit of allowing you to release updates and improvements outside of the GPDB release cycle and we can avoid the testing and environment challenges. Additionally that should allow us to side-step the need for a deeper legal review and while then the memory management issue would not need to be addressed, we’d still encourage you to do so.

Many thanks again.

**Comment by wjzeng** [comment]
Thank you very much for your reply.
Your suggestion is very helpful to me, I hope to have in-depth communication.

**Comment by Eulerizeit** [comment]
Given the comment above about us not being able to accept this PR I'm going to close it. 

Again, that you for the work you did.

**Pull request closed without merging**